cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/DatasetTypeTable.java

```diff
@@ -0,0 +1,257 @@
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Ranges;
+import com.google.gson.Gson;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Dataset types and modules metadata store
+ */
+public class DatasetTypeTable {
+
+  private static final Gson GSON = new Gson();
+
+  private final StructuredTable typeTable;
+  private final StructuredTable moduleTable;
+
+  public DatasetTypeTable(StructuredTable moduleTable, StructuredTable typeTable) {
+    this.moduleTable = moduleTable;
+    this.typeTable = typeTable;
+  }
+
+  public static DatasetTypeTable create(StructuredTableContext context) throws TableNotFoundException {
+    return new DatasetTypeTable(context.getTable(StoreDefinition.DatasetTypeStore.MODULE_TYPES),
+                                context.getTable(StoreDefinition.DatasetTypeStore.DATASET_TYPES));
+  }
+
+  /**
+   * Retrieves a module from the given namespace
+   *
+   * @param datasetModuleId the {@link DatasetModuleId} for the module to retrieve
+   * @return {@link DatasetModuleMeta} for the module if found in the specified namespace, null otherwise
+   */
+  @Nullable
+  public DatasetModuleMeta getModule(DatasetModuleId datasetModuleId) throws IOException {
+    return get(getModuleKey(datasetModuleId.getNamespace(), datasetModuleId.getEntityName()),
+               moduleTable, DatasetModuleMeta.class);
+  }
+
+  /**
+   * Tries to find a module in the specified namespace first. If it fails, tries to find it in the system namespace
+   *
+   * @param datasetModuleId {@link DatasetModuleId} for the module to retrieve
+   * @return {@link DatasetModuleMeta} for the module if found either in the specified namespace or in the system
+   * namespace, null otherwise
+   */
+  @Nullable
+  public DatasetModuleMeta getModuleWithFallback(DatasetModuleId datasetModuleId) throws IOException {
+    // Try to find module in the specified namespace first
+    DatasetModuleMeta moduleMeta = getModule(datasetModuleId);
+    // if not found, try to load it from system namespace
+    if (moduleMeta == null) {
+      moduleMeta = getModule(NamespaceId.SYSTEM.datasetModule(datasetModuleId.getEntityName()));
+    }
+    return moduleMeta;
+  }
+
+  @Nullable
+  public DatasetModuleMeta getModuleByType(DatasetTypeId datasetTypeId) throws IOException {
+    DatasetModuleId datasetModuleId =
+      get(getTypeKey(datasetTypeId.getNamespace(), datasetTypeId.getEntityName()), typeTable, DatasetModuleId.class);
+
+    if (datasetModuleId == null) {
+      return null;
+    }
+    // TODO: Slightly strange. Maybe change signature to accept NamespaceId separately from typeName
```

Comment:
not sure I understand this TODO

Reply:
I don't either - it's existing code. will delete